src/graphql/merge.js

```diff
@@ -0,0 +1,317 @@
+  const fragments = mergeFragments(queryAFragments, queryBFragments)
+  // Finally return the merged definitions and fragments (loc in the AST nodes is not important).
+  queryA.definitions = [
+    definition,
+    ...fragments
+  ]
+  return queryA
+}
+
+/**
+ * Merge two arrays of fragments, returning a single array of fragments.
+ *
+ * @param {Array<FragmentDefinitionNode>} fragmentsA
+ * @param {Array<FragmentDefinitionNode>} fragmentsB
+ * @return {Array<FragmentDefinitionNode>}
+ */
+function mergeFragments (fragmentsA, fragmentsB) {
+  const fragments = fragmentsA
+  /**
+   * @type {Object.<String, FragmentDefinitionNode>}
+   */
+  const fragmentsByTypeCondition = {}
+  fragments.forEach(fragment => {
+    fragmentsByTypeCondition[fragment.typeCondition.name.value] = fragment
+  })
+  fragmentsB.forEach(fragment => {
+    const typeConditionName = fragment.typeCondition.name.value
+    const existingFragment = fragmentsByTypeCondition[typeConditionName]
+    if (!existingFragment) {
+      // If it is a new fragment, then no need to worry and just add it to the array.
+      fragments.push(fragment)
+    } else {
+      // Else we need to merge the fragments.
+      existingFragment.selectionSet = mergeSelectionSets(existingFragment.selectionSet, fragment.selectionSet)
+      existingFragment.directives = mergeDirectives(existingFragment.directives, fragment.directives)
+    }
+  })
+  return fragments
+}
+
+/**
+ * Merge two selection sets, returning a single selection set.
+ *
+ * @param {SelectionSetNode} selectionSetA
+ * @param {SelectionSetNode} selectionSetB
+ * @return {SelectionSetNode|undefined}
+ */
+function mergeSelectionSets (selectionSetA, selectionSetB) {
+  if (!selectionSetA && !selectionSetB) {
+    return undefined
+  }
+  if ((!selectionSetA && selectionSetB) || (selectionSetA && !selectionSetB)) {
+    throw new Error('Selection sets must be either both undefined, or both defined')
+  }
+  const selectionSet = selectionSetA
+  /**
+   * @type {Object.<String, SelectionNode>}
+   */
+  const selectionsByName = {}
+  selectionSet.selections.forEach(selection => {
+    if (selection.kind === 'InlineFragment') {
+      throw new Error('Found a selection of type "InlineFragment". Only "Field" and "FragmentSpread" are supported')
+    }
+    const selectionName = selection.alias ? selection.alias.value : selection.name.value
+    selectionsByName[selectionName] = selection
+  })
+  selectionSetB.selections.forEach(field => {
+    if (field.kind === 'InlineFragment') {
+      throw new Error('Found a selection of type "InlineFragment". Only "Field" and "FragmentSpread" are supported')
+    }
+    const selectionName = field.alias ? field.alias.value : field.name.value
+    const existingField = selectionsByName[selectionName]
+    if (!existingField) {
+      // If it is a new field, then no need to worry and just add it to the array.
+      // TBD: is selections read-only?
```

Comment:
Ah, had to read the code to remember why I added this comment. It's not really read-only, since it's quite hard to make anything private in JS. It's just a note as I think we are not supposed to change it, but the guys in `graphql-js` might change their minds someday.

https://github.com/graphql/graphql-js/blob/d32c5a1348b6d3d0bbede7fe063c56976c1903c3/src/language/ast.ts#L294